From b24f428cf6e66c1acd518c2756f7f1cb6cd4c138 Mon Sep 17 00:00:00 2001
From: Thomas Wu <quic_wthomas@quicinc.com>
Date: Thu, 1 Aug 2024 00:41:45 -0700
Subject: [PATCH] wifi: ath12k: Add WARN_ON checks for nss value

NSS should not have a value of 0 and should always
have a value from 1 to 8.

To ensure this statement is true, WARN_ON checks
have been added to monitor this value.

Signed-off-by: Thomas Wu <quic_wthomas@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_tx.c | 2 ++
 drivers/net/wireless/ath/ath12k/mac.c   | 2 ++
 2 files changed, 4 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -997,6 +997,9 @@ void ath12k_dp_tx_update_txcompl(struct
 	else
 		arsta->txrate.nss = arsta->peer_nss;
 
+	WARN_ON_ONCE(arsta->txrate.nss < 1 ||
+		     (arsta->txrate.nss > hweight32(ar->pdev->cap.tx_chain_mask)));
+
 	if (pkt_type == HAL_TX_RATE_STATS_PKT_TYPE_11A ||
 	    pkt_type == HAL_TX_RATE_STATS_PKT_TYPE_11B) {
 		ret = ath12k_mac_hw_ratecode_to_legacy_rate(mcs,
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -4063,6 +4063,9 @@ static void ath12k_peer_assoc_prepare(st
 
 	arsta->peer_nss = arg->peer_nss;
 
+	WARN_ON_ONCE(arsta->peer_nss < 1 ||
+		     (arsta->peer_nss > hweight32(ar->pdev->cap.tx_chain_mask)));
+
 	/* TODO: amsdu_disable req? */
 }
 
